src/main/java/hudson/plugins/jira/pipeline/JiraJobAction.java

```diff
@@ -0,0 +1,124 @@
+package hudson.plugins.jira.pipeline;
+
+import com.google.common.base.Objects;
+import hudson.Extension;
+import hudson.model.Action;
+import hudson.model.Job;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import hudson.plugins.jira.JiraSite;
+import hudson.plugins.jira.model.JiraIssue;
+import jenkins.branch.MultiBranchProject;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.export.Exported;
+import org.kohsuke.stapler.export.ExportedBean;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.net.URL;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * JiraJobAction is to store a reference to the {@link JiraIssue} that represents work
+ * being done for a {@link WorkflowJob} (branch or PR) belonging to a {@link MultiBranchProject}
+ */
+@ExportedBean
+public class JiraJobAction implements Action {
+
+    private static final Logger LOGGER = Logger.getLogger(JiraJobAction.class.getName());
+
+    private final Job<?, ?> owner;
+    private final JiraIssue jiraIssue;
+
+    @DataBoundConstructor
+    public JiraJobAction(Job<?, ?> owner, JiraIssue jiraIssue) {
+        this.owner = owner;
+        this.jiraIssue = jiraIssue;
+    }
+
+    @Exported
+    public JiraIssue getJiraIssue() {
+        return jiraIssue;
+    }
+
+    @Exported
+    public String getServerURL() {
+        JiraSite jiraSite = JiraSite.get(owner);
+        URL url = jiraSite != null ? Objects.firstNonNull(jiraSite.url, jiraSite.alternativeUrl) : null;
+        return url != null ? url.toString() : null;
+    }
+
+    /**
+     * Adds a {@link JiraJobAction} to a {@link WorkflowJob} if it belongs to a {@link MultiBranchProject}
+     * and its name contains an JIRA issue key
+     * @param job to add the property to
+     * @param site to fetch issue data
+     * @throws IOException if something goes wrong fetching the JIRA issue
+     */
+    public static void setAction(@Nonnull Job<?, ?> job, @Nonnull JiraSite site) throws IOException {
+        // If there is already a action set then skip
+        if (job.getAction(JiraJobAction.class) != null) {
```

Comment:
`setAction` is called on every run. If the fetch to JIRA fails, then for the next run we have the opportunity to fetch again. Eventually the data is stored on the action and this branch is executed.